Revamp ytt schema (and Data Values) documentation

pivotaljohn opened the pull request:
An insight from usability testing is that documentation around Data Values and Schema were fragmented and (inadvertently) buried important information.

This effort attempts to incorporate that insight in two ways:
1. fold in the use of schema as an aspect of how to use data values
2. separate guidance from reference

gcheadle-vmware committed:
Rename @schema/match data_values=True to @data/values-schema

jtigger committed:
Revamp ytt Data Values docs to include Schema

- rename "Techniques" to "How To" to more readily communicate contained
  pages are guides.
- in guides, weave in usage of Schema as a step of using Data Values as
  this is how users likely think of it: a way to declare Data Values.
- remove mention of the `--enable-experiment-schema` flag.
- extract definitions into a reference from instructions (guides).
- various reformatting to improve readability.
- relocate "Schema module" from "@ytt library" to "Reference" as modules
  ought to be limited to files that template authors `load()`.
- adjust references to renamed files and anchors.

gcheadle-vmware reviewed (commented):
Left some comments, most are small easy changes. Overall it looks and feels like an improvement!

One note about a file that is not included in this change set: the ytt Documentation Homepage 'About ytt'.  This doc page introduces templating and data values, then links to the "Data Values" Reference page. I'm wondering if you think this is sufficient, or if it would be friendlier to link to "Use Data Values" page?

pivotaljohn commented:
> I'm wondering if you think this is sufficient, or if it would be friendlier to link to "Use Data Values" page?

Absolutely! Nice catch... doin' it...

pivotaljohn commented on content/ytt/docs/latest/how-to-use-data-values.md:
```diff
@@ -0,0 +1,98 @@
+- [How To Write Schema](how-to-write-schema.md) guide
+- [Data Values](ytt-data-values.md) reference
+- [Data Values Schema](lang-ref-ytt-schema.md) reference
+- Schema Migration Guide (TBD)
```
Thank you!

pivotaljohn commented on content/ytt/docs/latest/how-to-write-schema.md:
````diff
@@ -0,0 +1,217 @@
+system_domain: ""
+```
+
+declares the Data Value, with the name `system_domain` to be of type **string** with the default value of an empty string (i.e. `""`).
````
nice touch. I bolded "empty string" but I think the code-style treatment of `system_domain` is enough?!?

pivotaljohn commented on content/ytt/docs/latest/how-to-write-schema.md:
```diff
@@ -0,0 +1,217 @@
+
+However, as described in [Data Values Schema Reference: Defaults for Arrays](lang-ref-ytt-schema.md#defaults-for-arrays), the two arrays, `app_domains` and `databases` are each an empty list (i.e. `[]`) by default.
+
+That said, when an item is added to `database`, each item will...
```
Yeah, understandable that it's not understandable... reworking. 👍🏻

pivotaljohn commented on content/ytt/docs/latest/how-to-write-schema.md:
```diff
@@ -0,0 +1,217 @@
+
+
+---
+## Next Steps
```
oops!  Meant to flow back to the main "Using Data Values" page... will finish/fix.

pivotaljohn commented on content/ytt/docs/latest/ytt-data-values.md:
````diff
@@ -112,68 +197,12 @@ fifth: new-val5
-```
 
-## Data values order
+### Data Values merge order
````
sweet! Thanks.

jtigger committed:
Address feedback from Garrett

- from the tool introduction, link to Data Values guide rather than
  reference for a friendlier introduction.
- elaborate (and be more precise) about how maps inside arrays are
  defaulted.
- actually provide a next step. :grimace:

cari-lynn commented on content/ytt/docs/latest/how-to-write-schema.md:
```diff
@@ -0,0 +1,238 @@
+- `int_or_string` is, by default, an empty string
+- it can accept an **integer** or a **string** ... or any other type.
+
+If it is critical to ensure that the type of `int_or_string` to be _only_ an integer or string, one can include a validating library that does so explicitly:
```
This is a great tip. Providing a way around a feature schema doesn't yet support makes these docs feel really complete.

cari-lynn commented on content/ytt/docs/latest/how-to-write-schema.md:
````diff
@@ -0,0 +1,238 @@
+end
+```
+
+### Declaring "Pass-through" Data Values
````
I read this under the headings Setting Default Values > Specific Use Cases, and expected to learn how to solve this use case in this paragraph but I only see a statement of the use case. I think the answer for it is to use `schema/type any=True`, It may be useful to say that here.

pivotaljohn replied:
Yup, absolutely forgot to flesh out that use-case. Great catch, @cari-lynn!

Will complete.

cari-lynn reviewed (approved):
Really excited to see these changes. Thank you for all your effort on these!

jtigger committed:
Flesh out missed schema usage scenario
